### PR TITLE
feat(New hiseqx validation functions)

### PR DESCRIPTION
**Pull request by karlnyr**
This PR adds new functions in validation of hiseqx SampleSheet. Specifically, they check wether there are more than type of indexes present(i.e., single or dual indexes) as well as an implementation of a sanity length check on dual indexes 

**How to prepare for test**:
- [x] ssh to hasta
- [x] install on stage:
`bash servers/resources/hasta.scilifelab.se/update-demux-stage.sh feat/hiseq-samplesheet-validation`

**How to test**:
- [x] login to hasta
- [x] `us`
- [x] run `demux sheet validate --application wgs` on:
    - [x] samplesheet with dual indexes of variable length
    - [x] samplesheet with multiple types of indexes

**Expected test outcome**:
- [x] SampleSheets raises validation errors 
- [x] Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [x] tests executed by @karlnyr 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by henrikstranneheim** [comment]
@karlnyr What is the status of this one?

**Comment by karlnyr** [comment]
> @karlnyr What is the status of this one?

I've written most of the tests already but got interrupted with production during Christmas. I Will most likely pick this up again next week 👍

**Comment by karlnyr** [comment]
I mixed up this PR and another one. I have now finished up the function and added a test for the function. The function is to be used in the demultiplexing workflow of hiseqx, so testing on hasta could be performed to test it.

**Comment by karlnyr** [comment]
Samplesheet with multiple index types:
image

Test:
image

Samplesheet with variable dual index length:
image

Test:
image

**Comment by karlnyr** [comment]
@barrystokman here it is

@henrikstranneheim only 5 months later after your first ping on the repo 👍

**Comment by karlnyr** [comment]
@barrystokman can you give me a review just to make it official? You may have already looked through the code but.... for appearance sake 💯

**Review by barrystokman — CHANGES_REQUESTED** [review]
Looks good, just had some minor remarks.

**Comment by barrystokman** [comment]
Let's decide what to do with the 2500 tests since I'm doing those in my PR as well.

File: demux/utils/samplesheet.py [hunk 1]
```diff
@@ -717,3 +751,10 @@ def to_demux(self, delim=",", end="\n"):
         for line in data_lines:
             rs.append(delim.join(line))
         return end.join(rs)
+
+
+if __name__ == "__main__":
+    sample_sheet = HiSeqXSamplesheet(
+        "/tests/fixtures/hiseqx/x_samplesheet_multiple_index.csv"
+    )
+    sample_sheet.validate()
```
[the remark this answers is not in the document]
Comment (karlnyr) [reply demux/utils/samplesheet.py:760]: hehe, oopsie 😅

File: demux/utils/samplesheet.py [hunk 2]
```diff
@@ -18,7 +22,7 @@ def __str__(self):
         )
 
 
-class SampleSheetParseException(Exception):
+class SampleSheetParsexception(Exception):
```
Comment (barrystokman) [inline demux/utils/samplesheet.py:25]:
```suggestion
class SampleSheetParseException(Exception):
```

File: demux/utils/samplesheet.py [hunk 3]
```diff
@@ -120,7 +124,7 @@ def _get_header_key(self, key):
     def parse(self, samplesheet_path, delim=","):
         """
         Parses a Samplesheet, with their fake csv format.
-        Should be instantiated with the samplesheet path as an argument.
+        Should be instancied with the samplesheet path as an argument.
```
Comment (barrystokman) [inline demux/utils/samplesheet.py:127]:
```suggestion
        Should be instantiated with the samplesheet path as an argument.
```

File: demux/utils/samplesheet.py [hunk 4]
```diff
@@ -142,7 +146,7 @@ def parse(self, samplesheet_path, delim=","):
                 self.section[name].append(line)
 
         if self.DATA not in self.section:
-            raise SampleSheetParseException("No data found!")
+            raise SampleSheetParsexception("No data found!")
```
Comment (barrystokman) [inline demux/utils/samplesheet.py:149]:
```suggestion
            raise SampleSheetParseException("No data found!")
```

File: demux/utils/samplesheet.py [hunk 5]
```diff
@@ -222,7 +226,7 @@ def is_pooled_lane(self, lane, column="lane"):
         return False
 
     def is_pooled_lane_r(self, lane, column="lane"):
-        """ Return True if lane contains multiple samples based on the original header """
+        """ Return True if lane contains multiple samples based on the orignal header """
```
Comment (barrystokman) [inline demux/utils/samplesheet.py:229]:
```suggestion
        """ Return True if lane contains multiple samples based on the original header """
```

File: demux/utils/samplesheet.py [hunk 6]
```diff
@@ -248,7 +252,7 @@ def _validate_length(section):
                         return (msg, i + 2)
             return True
 
-        def _validate_unique_index(samplesheet):
+        def _validate_uniq_index(samplesheet):
```
Comment (barrystokman) [inline demux/utils/samplesheet.py:255]:
```suggestion
        def _validate_unique_index(samplesheet):
```

File: demux/utils/samplesheet.py [hunk 7]
```diff
@@ -258,13 +262,12 @@ def _validate_unique_index(samplesheet):
                         if index not in sample_of:
                             sample_of[index] = set()
                         sample_of[index].add(line["sample_id"])
+
                     for index, samples in sample_of.items():
                         if len(samples) > 1:
-                            samples_list = list(samples)
-                            samples_list.sort()
```
Comment (barrystokman) [inline demux/utils/samplesheet.py:264]: This was in my PR and need to be in place. You removed stuff that was supposed to be in there!

File: demux/utils/samplesheet.py [hunk 8]
```diff
@@ -258,13 +262,12 @@ def _validate_unique_index(samplesheet):
...
                             return (
                                 "Same index for {} on lane {}".format(
-                                    " , ".join(samples_list), lane
```
Comment (barrystokman) [inline demux/utils/samplesheet.py:267]: Please leave this in.

File: demux/utils/samplesheet.py [hunk 9]
```diff
@@ -282,7 +285,7 @@ def _validate_sample_name(samplesheet):
                         i + 2,
                     )
 
-        rs = _validate_unique_index(self.samplesheet)
+        rs = _validate_uniq_index(self.samplesheet)
```
Comment (barrystokman) [inline demux/utils/samplesheet.py:288]:
```suggestion
        rs = _validate_unique_index(self.samplesheet)
```

Maybe it's best to revert some stuff, this was all put in by me, very much intentionally.

File: tests/conftest.py [hunk 10]
```diff
@@ -96,10 +89,10 @@ def fixture_hiseqx_samplesheet(hiseqx_dir: Path) -> Path:
     return hiseqx_dir / "x_samplesheet.csv"
 
 
-@pytest.fixture(name="x_faulty_samplesheet_path")
-def fixture_x_faulty_samplesheet(hiseqx_dir: Path) -> Path:
+@pytest.fixture(name="x_samplesheet_wrong_font_path")
```
Comment (barrystokman) [inline tests/conftest.py:92]: "wrong font" was a joke

File: tests/conftest.py [hunk 11]
```diff
@@ -151,8 +134,8 @@ def fixture_novaseq_runparameters_file(novaseq_runs_dir: Path) -> Dict[str, Path
     }
 
 
-@pytest.fixture(scope="function", name="novaseq_runparameters_api")
-def fixture_novaseq_runparameters_api(novaseq_runs_dir: Path):
+@pytest.yield_fixture(scope="function", name="novaseq_runparameters_api")
```
Comment (barrystokman) [inline tests/conftest.py:137]: this needs to be put back

**Review by barrystokman — CHANGES_REQUESTED** [review]
You removed stuff that was supposed to be in there, things I added in my PR. Maybe it's best to revert some of the merge stuff.

**Comment by karlnyr** [comment]
@barrystokman I think I finally solved it all. After reverting, messing up merging 60+ times and pushing I think I finally managed to get the master into this branch. Can you please double check and then I will merge this and get on with the next PR

**Review by barrystokman — APPROVED** [review]
🏁

**Comment by henrikstranneheim** [comment]
Well done! 🎉

**Pull request merged**